File: obsidian-mcp-ts/src/obsidian/advanced_search.ts
import fs from "fs";
import path from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getVaultPath } from "./vault.js";
import { findMarkdownFiles, relPath, stem, safePath } from "./utils.js";

function fuzzyMatch(query: string, text: string): boolean {
  const qLower = query.toLowerCase();
  const tLower = text.toLowerCase();
  let qi = 0;
  for (const ch of tLower) {
    if (qi < qLower.length && ch === qLower[qi]) qi++;
  }
  return qi === qLower.length;
}

export function registerAdvancedSearchTools(server: McpServer) {
  server.tool(
    "search_by_regex",
    "Search for notes using regular expression pattern",
    {
      pattern: z.string().describe("Regular expression pattern to search for"),
      folder: z.string().optional().describe("Optional folder path to limit search"),
    },
    async ({ pattern, folder }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? safePath(vaultPath, folder) : vaultPath;
        if (!fs.existsSync(searchPath)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder}`, matches: [] }) }] };

        let regex: RegExp;
        try {
          regex = new RegExp(pattern, "gim");
        } catch (e: any) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Invalid regex pattern: ${e.message}`, matches: [] }) }] };
        }

        const matches: any[] = [];
        for (const mdFile of findMarkdownFiles(searchPath)) {
          try {
            const content = fs.readFileSync(mdFile, "utf-8");
            const matchList: any[] = [];
            let m: RegExpExecArray | null;
            while ((m = regex.exec(content)) !== null) {
              matchList.push({
                line: content.slice(0, m.index).split("\n").length,
                match: m[0],
                start: m.index,
                end: m.index + m[0].length,
              });
            }
            if (matchList.length > 0) {
              matches.push({
                name: stem(mdFile),
                path: relPath(vaultPath, mdFile),
                matches: matchList.slice(0, 10),
                total_matches: matchList.length,
              });
            }
          } catch { continue; }
        }

        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ pattern, folder: folder || "root", total_files_matched: matches.length, matches }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message, matches: [] }) }] };
      }
    }
  );

  server.tool(
    "search_by_date_range",
    "Search for notes created or modified within a date range",
    {
      start_date: z.string().describe("Start date in YYYY-MM-DD format"),
      end_date: z.string().optional().describe("End date in YYYY-MM-DD format (defaults to today)"),
      folder: z.string().optional().describe("Optional folder path to limit search"),
    },
    async ({ start_date, end_date, folder }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? safePath(vaultPath, folder) : vaultPath;
        if (!fs.existsSync(searchPath)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder}`, matches: [] }) }] };

        const start = new Date(start_date + "T00:00:00");
        const end = end_date ? new Date(end_date + "T23:59:59") : new Date();
        if (isNaN(start.getTime()) || isNaN(end.getTime())) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: "Invalid date format. Use YYYY-MM-DD", matches: [] }) }] };
        }

        const matches: any[] = [];
        for (const mdFile of findMarkdownFiles(searchPath)) {
          try {
            const stat = fs.statSync(mdFile);
            const modified = stat.mtime;
            const created = stat.birthtime;
            if ((modified >= start && modified <= end) || (created >= start && created <= end)) {
              matches.push({
                name: stem(mdFile),
                path: relPath(vaultPath, mdFile),
                created: created.toISOString().replace("T", " ").slice(0, 19),
                modified: modified.toISOString().replace("T", " ").slice(0, 19),
              });
            }
          } catch { continue; }
        }

        matches.sort((a, b) => b.modified.localeCompare(a.modified));
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ start_date, end_date: end_date || new Date().toISOString().slice(0, 10), folder: folder || "root", total_matches: matches.length, matches }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message, matches: [] }) }] };
      }
    }
  );

  server.tool(
    "fuzzy_search_note",
    "Fuzzy search for notes by name",
    {
      query: z.string().describe("Search query (characters in order)"),
      folder: z.string().optional().describe("Optional folder path to limit search"),
      limit: z.number().default(20).describe("Maximum number of results to return"),
    },
    async ({ query, folder, limit }) => {
      try {
        const vaultPath = getVaultPath();
        const searchPath = folder ? safePath(vaultPath, folder) : vaultPath;
        if (!fs.existsSync(searchPath)) return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Folder not found: ${folder}`, matches: [] }) }] };

        const matches: any[] = [];
        for (const mdFile of findMarkdownFiles(searchPath)) {
          const noteName = stem(mdFile);
          if (fuzzyMatch(query, noteName)) {
            matches.push({ name: noteName, path: relPath(vaultPath, mdFile) });
            if (matches.length >= limit) break;
          }
        }

        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ query, folder: folder || "root", total_matches: matches.length, matches }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message, matches: [] }) }] };
      }
    }
  );
}

File: obsidian-mcp-py/obsidian/advanced_search.py
"""
Obsidian Advanced Search

Provides advanced search capabilities for Obsidian notes.
"""

import re
from pathlib import Path
from typing import Optional, List, Dict
from datetime import datetime
from obsidian.vault import get_vault_path, safe_path


def fuzzy_match(query: str, text: str) -> bool:
    """
    Simple fuzzy matching - checks if query characters appear in order in text.
    
    Args:
        query: Search query
        text: Text to search in
        
    Returns:
        True if fuzzy match found
    """
    query_lower = query.lower()
    text_lower = text.lower()
    
    query_idx = 0
    for char in text_lower:
        if query_idx < len(query_lower) and char == query_lower[query_idx]:
            query_idx += 1
    
    return query_idx == len(query_lower)


def register_advanced_search_tools(mcp):
    """
    Register all advanced search-related MCP tools.
    
    Args:
        mcp: FastMCP instance to register tools with
    """
    
    @mcp.tool()
    def search_by_regex(pattern: str, folder: Optional[str] = None) -> dict:
        """
        Search for notes using regular expression pattern.
        
        Args:
            pattern: Regular expression pattern to search for
            folder: Optional folder path to limit search
        
        Returns:
            Dictionary with matching notes and matches
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "matches": []}
            
            matches = []
            
            try:
                regex = re.compile(pattern, re.IGNORECASE | re.MULTILINE)
            except re.error as e:
                return {"error": f"Invalid regex pattern: {str(e)}", "matches": []}
            
            for md_file in search_path.rglob("*.md"):
                if ".obsidian" in md_file.parts:
                    continue
                
                try:
                    content = md_file.read_text(encoding="utf-8")
                    regex_matches = regex.finditer(content)
                    
                    match_list = []
                    for match in regex_matches:
                        match_list.append({
                            "line": content[:match.start()].count('\n') + 1,
                            "match": match.group(0),
                            "start": match.start(),
                            "end": match.end()
                        })
                    
                    if match_list:
                        rel_path = md_file.relative_to(vault_path)
                        matches.append({
                            "name": md_file.stem,
                            "path": str(rel_path),
                            "matches": match_list[:10],  # Limit to first 10 matches per file
                            "total_matches": len(match_list)
                        })
                except Exception:
                    continue
            
            return {
                "pattern": pattern,
                "folder": folder or "root",
                "total_files_matched": len(matches),
                "matches": matches
            }
        except Exception as e:
            return {"error": str(e), "matches": []}
    
    
    @mcp.tool()
    def search_by_date_range(start_date: str, end_date: Optional[str] = None, folder: Optional[str] = None) -> dict:
        """
        Search for notes created or modified within a date range.
        
        Args:
            start_date: Start date in YYYY-MM-DD format
            end_date: End date in YYYY-MM-DD format (defaults to today)
            folder: Optional folder path to limit search
        
        Returns:
            Dictionary with notes in date range
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "matches": []}
            
            try:
                start = datetime.strptime(start_date, "%Y-%m-%d")
                if end_date:
                    end = datetime.strptime(end_date, "%Y-%m-%d")
                else:
                    end = datetime.now()
            except ValueError as e:
                return {"error": f"Invalid date format: {str(e)}. Use YYYY-MM-DD", "matches": []}
            
            matches = []
            
            for md_file in search_path.rglob("*.md"):
                if ".obsidian" in md_file.parts:
                    continue
                
                try:
                    stat = md_file.stat()
                    modified_time = datetime.fromtimestamp(stat.st_mtime)
                    created_time = datetime.fromtimestamp(stat.st_ctime)
                    
                    # Check if modified or created in range
                    if start <= modified_time <= end or start <= created_time <= end:
                        rel_path = md_file.relative_to(vault_path)
                        matches.append({
                            "name": md_file.stem,
                            "path": str(rel_path),
                            "created": created_time.strftime("%Y-%m-%d %H:%M:%S"),
                            "modified": modified_time.strftime("%Y-%m-%d %H:%M:%S")
                        })
                except Exception:
                    continue
            
            matches.sort(key=lambda x: x["modified"], reverse=True)
            
            return {
                "start_date": start_date,
                "end_date": end_date or datetime.now().strftime("%Y-%m-%d"),
                "folder": folder or "root",
                "total_matches": len(matches),
                "matches": matches
            }
        except Exception as e:
            return {"error": str(e), "matches": []}
    
    
    @mcp.tool()
    def fuzzy_search_note(query: str, folder: Optional[str] = None, limit: int = 20) -> dict:
        """
        Fuzzy search for notes by name.
        
        Args:
            query: Search query (characters in order)
            folder: Optional folder path to limit search
            limit: Maximum number of results to return
        
        Returns:
            Dictionary with matching note names
        """
        try:
            vault_path = get_vault_path()
            
            if folder:
                search_path = safe_path(vault_path, folder)
            else:
                search_path = vault_path
            
            if not search_path.exists():
                return {"error": f"Folder not found: {folder}", "matches": []}
            
            matches = []
            
            for md_file in search_path.rglob("*.md"):
                if ".obsidian" in md_file.parts:
                    continue
                
                note_name = md_file.stem
                if fuzzy_match(query, note_name):
                    rel_path = md_file.relative_to(vault_path)
                    matches.append({
                        "name": note_name,
                        "path": str(rel_path)
                    })
                    
                    if len(matches) >= limit:
                        break
            
            return {
                "query": query,
                "folder": folder or "root",
                "total_matches": len(matches),
                "matches": matches
            }
        except Exception as e:
            return {"error": str(e), "matches": []}
    
    